xdg-shell: dont overwrite for_window scratchpad geometry

sahinf opened the pull request:
fixes #8493, fixes #8018, fixes #8711, fixes #8749, fixes #9127

sahinf commented:
Non-terminal windows should also get matched. lmk if that's not the case (it is for me).

emersion commented on sway/criteria.c:
```diff
@@ -419,12 +419,6 @@ static bool criteria_matches_view(struct criteria *criteria,
 	}
 #endif
 
-	if (criteria->floating) {
-		if (!container_is_floating(view->container)) {
-			return false;
-		}
-	}
```
I don't understand this change. Doesn't this result in `criteria->floating` being always ignored and floating criteria matching tiled containers?

emersion commented on sway/desktop/xdg_shell.c:
```diff
@@ -321,7 +321,7 @@ static void handle_commit(struct wl_listener *listener, void *data) {
 		// we only recenter the surface.
 		memcpy(&view->geometry, new_geo, sizeof(struct wlr_box));
-		if (container_is_floating(view->container)) {
+		if (container_is_floating(view->container) && !container_is_scratchpad_hidden(view->container)) {
```
Doesn't that result in a broken view size if a view in the scratchpad resizes itself?

sahinf replied:
It didn't always but sometimes yeah. Just checking the node instruction and serial instead fixed all 3 issues tbh.

sahinf committed:
xdg_shell: don't overwrite for_window directive geometry

When a floating container is resized (e.g., via for_window rules or
internal transactions), a race condition occurs if the client commits a
buffer with its old geometry before processing the new configure event.

Sway currently accepts these stale dimensions and calls
view_update_size, causing the container to snap back to the client's
previous size and overwriting the compositor's intended geometry.

sahinf committed:
xdg_shell: don't overwrite for_window directive geometry

When a floating container is resized (e.g., via for_window rules or
internal transactions), a race condition occurs if the client commits a
buffer with its old geometry before processing the new configure event.

Sway currently accepts these stale dimensions and calls
view_update_size, causing the container to snap back to the client's
previous size and overwriting the compositor's intended geometry.

sahinf commented:
8927.in
```swayconfig
bindsym i scratchpad show
exec foot --app-id=1
for_window [app_id="1"] floating enable, resize set 300 px 300 px
exec foot --app-id=2
for_window [app_id="2"] floating enable, resize set 2400 1000, move scratchpad
exec foot --app-id=3
for_window [app_id="3"] floating enable, resize set 80 ppt 70 ppt, move scratchpad
```
8927.gdb
```gdb
break sway/desktop/xdg_shell.c:333
commands 
    printf "Acked serial: %u\n", xdg_surface->current.configure_serial
    print *xdg_surface
    continue
end

break root_scratchpad_add_container
commands
    printf "con=%p, width=%d, height=%d, floating=%d\n", con, con->pending.width, con->pending.height, container_is_floating(con)
    continue
end
break container_floating_set_default_size
commands
    printf "Before default: width=%d, height=%d\n", con->pending.width, con->pending.height
    continue
end
break configure if view->app_id && strcmp(view->app_id, "scratch3") == 0
commands
    printf "Sway requests: %dx%d\n", width, height
    continue
end
break get_constraints
commands
    printf "min: %.0fx%.0f, max: %.0fx%.0f\n", *min_width, *min_height, *max_width, *max_height
    continue
end
break floating_fix_coordinates
commands
    printf "con=%p, width=%d, height=%d, old_box=(%d,%d %dx%d), new_box=(%d,%d %dx%d)\n", \
        con, con->pending.width, con->pending.height, \
        old->x, old->y, old->width, old->height, \
        new->x, new->y, new->width, new->height
    continue
end
break container_floating_move_to_center
commands
    printf "Before center: width=%d, height=%d\n", con->pending.width, con->pending.height
    continue
end

run --config 8927.in
```

```sh
$ meson compile -C debug && WLR_RENDERER=pixman gdb -x 8927.gdb -q ./debug/sway/sway
press i
```

emersion commented on sway/desktop/xdg_shell.c:
```diff
@@ -321,7 +321,8 @@ static void handle_commit(struct wl_listener *listener, void *data) {
 		memcpy(&view->geometry, new_geo, sizeof(struct wlr_box));
-		if (container_is_floating(view->container)) {
+		bool pending_configures = xdg_surface->current.configure_serial < xdg_surface->scheduled_serial;
+		if (container_is_floating(view->container) && !view->container->node.instruction && !pending_configures) {
```
Are these two conditions necessary? If `pending_configures` is false, I was under the impression that the view couldn't have any pending instructions?

sahinf replied:
Based on my testing, node.instruction addresses whether a transaction is in progress. It is set during the resize and cleared when the transaction commits. So any client commit during this phase should be ignored to prevent overwriting the pending size, if that's the case. 

`pending_configures` is for when after the transaction commits, the instruction is cleared and the compositor sends a configure event with the new size but hte client may not have processed it yet. The client can still commit a buffer with its old size and `pending_configures` will be true because the `current.configure_serial` (last acked) is less than `scheduled_serial`. 

This was just observed/deduced behavior though and may not be intended deterministic effects, I guess.

emersion commented on sway/desktop/xdg_shell.c:
```diff
@@ -321,7 +321,8 @@ static void handle_commit(struct wl_listener *listener, void *data) {
 		// we only recenter the surface.
 		memcpy(&view->geometry, new_geo, sizeof(struct wlr_box));
-		if (container_is_floating(view->container)) {
+		bool pending_configures = xdg_surface->current.configure_serial < xdg_surface->scheduled_serial;
```
The configure serial can overflow the 32-bit integer. I think `xdg_surface->current.configure_serial != xdg_surface->scheduled_serial` should be safer?

emersion reviewed (commented):
I'm not exactly sure about the correctness of this fix… I don't think I fully grasp why we should ignore size changes when we have a pending configure.

If the check is applied to the inner `if` guarding `wlr_xdg_toplevel_set_size()` instead of the outer one, does the fix still work?

smlx commented on sway/desktop/xdg_shell.c:
```diff
@@ -321,7 +321,8 @@ static void handle_commit(struct wl_listener *listener, void *data) {
 		// we only recenter the surface.
 		memcpy(&view->geometry, new_geo, sizeof(struct wlr_box));
-		if (container_is_floating(view->container)) {
+		bool pending_configures = xdg_surface->current.configure_serial != xdg_surface->scheduled_serial;
+		if (container_is_floating(view->container) && !view->container->node.instruction && !pending_configures) {
```
Is the separate variable required? Maybe it is easier to read like this?

```suggestion
		if (container_is_floating(view->container) &&
				!view->container->node.instruction &&
				xdg_surface->current.configure_serial == xdg_surface->scheduled_serial) {
```

sahinf replied:
I would say it's easier to read with the variable but we can def split up that 2nd line for exceeding the 80 char limit.

smlx reviewed (commented):
Am I understanding this patch correctly?

1. Scratchpad windows (not visible) don't get instructions attached to them which is why you need to check that the scheduled and configure serials match (so you don't update the view size based on a client's stale buffer size).
1. Visible windows that have an instruction are already part of a committed transaction so it is logically incorrect to update their size and start a new commit with `transaction_commit_dirty_client()`?

Is that accurate?

sahinf commented:
> Am I understanding this patch correctly?
> 
> 1. Scratchpad windows (not visible) don't get instructions attached to them which is why you need to check that the scheduled and configure serials match (so you don't update the view size based on a client's stale buffer size).
> 2. Visible windows that have an instruction are already part of a committed transaction so it is logically incorrect to update their size and start a new commit with `transaction_commit_dirty_client()`?
> 
> Is that accurate?

1. It's general for all floating windows, not specific to scratchpad. The issue is for all windows with for_window directive
2. Yes that's my experience so far
